### PR TITLE
[SPARK-33861][SQL] Simplify conditional in predicate

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr simplify conditional in predicate, after this change we can push down the filter to datasource:

Expression | After simplify
-- | --
IF(cond, trueVal, false)                   | AND(cond, trueVal)
IF(cond, trueVal, true)                    | OR(NOT(cond), trueVal)
IF(cond, false, falseVal)                  | AND(NOT(cond), elseVal)
IF(cond, true, falseVal)                   | OR(cond, elseVal)
CASE WHEN cond THEN trueVal ELSE false END | AND(cond, trueVal)
CASE WHEN cond THEN trueVal END            | AND(cond, trueVal)
CASE WHEN cond THEN trueVal ELSE null END  | AND(cond, trueVal)
CASE WHEN cond THEN trueVal ELSE true END  | OR(NOT(cond), trueVal)
CASE WHEN cond THEN false ELSE elseVal END | AND(NOT(cond), elseVal)
CASE WHEN cond THEN false END              | false
CASE WHEN cond THEN true ELSE elseVal END  | OR(cond, elseVal)
CASE WHEN cond THEN true END               | cond

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -570,6 +571,33 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
...
+      val (whenVal, thenVal) = branches.head
+      And(whenVal, thenVal)
+    case i @ If(pred, trueVal, FalseLiteral) if i.dataType == BooleanType =>
+      And(pred, trueVal)
+    case e if e.dataType == BooleanType =>
+      e
+    case e =>
+      val message = "Expected a Boolean type expression in simplifyConditional, " +
+        s"but got the type `${e.dataType.catalogString}` in `${e.sql}`."
+      if (Utils.isTesting) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:591]: In this case, shall we raise exception always?

**Comment by wangyum** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:591]: I plan merge this logic to `ReplaceNullWithFalseInPredicate` in future commit: 
https://github.com/apache/spark/blob/09edff51ad5bce7b6f2b4b982d6df550a37d502e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala#L103-L111

**Comment by wangyum** [comment]
```
20:07:47.216 WARN org.apache.spark.sql.hive.execution.HiveCompatibilitySuite: 
=== Metrics of Analyzer/Optimizer Rules ===
Total number of runs: 1941616
Total time: 109.555372588 seconds

Rule                                                                                               Effective Time / Total Time                     Effective Runs / Total Runs                    

org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations                                   43227892210 / 43335027817                       3733 / 13640                                   
org.apache.spark.sql.hive.execution.PruneHiveTablePartitions                                       3991267073 / 4044441943                         106 / 5311                                     
org.apache.spark.sql.catalyst.optimizer.ColumnPruning                                              535610196 / 3853871083                          943 / 23531                                    
org.apache.spark.sql.catalyst.optimizer.CollapseProject                                            152456913 / 1700100541                          716 / 18220                                    
org.apache.spark.sql.catalyst.optimizer.PushDownPredicates                                         303440507 / 1439470162                          708 / 23531                                    
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveReferences                                  1004920252 / 1260007187                         2100 / 13640                                   
org.apache.spark.sql.catalyst.optimizer.RemoveNoopOperators                                        98769293 / 1168026956                           1036 / 23531                                   
org.apache.spark.sql.catalyst.optimizer.SimplifyConditionals                                       2264929 / 1032808041                            12 / 12909                                     
org.apache.spark.sql.catalyst.optimizer.BooleanSimplification                                      13648349 / 1025188033                           28 / 12909                                     
org.apache.spark.sql.catalyst.optimizer.SimplifyCasts                                              64831089 / 999273775                            204 / 12909                                    
org.apache.spark.sql.catalyst.optimizer.PruneFilters                                               3360581 / 974356833                             22 / 18220                                     
org.apache.spark.sql.catalyst.optimizer.OptimizeJsonExprs                                          0 / 939538541                                   0 / 12909                                      
org.apache.spark.sql.catalyst.optimizer.Optimizer$OptimizeSubqueries                               299198684 / 930080877                           18 / 10622                                     
org.apache.spark.sql.catalyst.optimizer.LikeSimplification                                         787725 / 928764647                              5 / 12909                                      
org.apache.spark.sql.catalyst.optimizer.CombineUnions                                              7340648 / 919100647                             30 / 18220                                     
org.apache.spark.sql.catalyst.optimizer.RemoveDispensableExpressions                               0 / 917768846                                   0 / 12909                                      
org.apache.spark.sql.catalyst.optimizer.NullPropagation                                            15369126 / 876020266                            72 / 12909                                     
org.apache.spark.sql.catalyst.optimizer.ConstantFolding                                            121197722 / 871577442                           406 / 12909                                    
org.apache.spark.sql.catalyst.optimizer.ReorderAssociativeOperator                                 1492835 / 829515880                             1 / 12909                                      
org.apache.spark.sql.catalyst.optimizer.SimplifyConditionalInPredicate                             598996 / 815043879                              1 / 12909                                      

```

**Comment by wangyum** [comment]
```
21:47:51.449 WARN org.apache.spark.sql.TPCDSQuerySuite: 
=== Metrics of Analyzer/Optimizer Rules ===
Total number of runs: 246192
Total time: 105.655380705 seconds

Rule                                                                              Effective Time / Total Time                     Effective Runs / Total Runs                    

org.apache.spark.sql.catalyst.optimizer.Optimizer$OptimizeSubqueries              13967311221 / 17930421970                       47 / 772                                       
org.apache.spark.sql.catalyst.optimizer.ColumnPruning                             1386295449 / 6773304521                         328 / 2383                                     
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveAggregateFunctions         1333101651 / 3589200583                         49 / 2166                                      
org.apache.spark.sql.catalyst.analysis.Analyzer$AddMetadataColumns                0 / 3410575751                                  0 / 2176                                       
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveSubquery                   2926377387 / 3067767576                         51 / 2166                                      
org.apache.spark.sql.catalyst.optimizer.PruneFilters                              32468139 / 2319485483                           5 / 1997                                       
org.apache.spark.sql.catalyst.analysis.DecimalPrecision                           1921173686 / 2150344481                         361 / 2166                                     
org.apache.spark.sql.catalyst.optimizer.PushDownPredicates                        884285005 / 1843322102                          767 / 2390                                     
org.apache.spark.sql.catalyst.optimizer.BooleanSimplification                     7784243 / 1672676489                            4 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.ReorderJoin                               708427120 / 1662902661                          177 / 1611                                     
org.apache.spark.sql.catalyst.optimizer.RemoveNoopOperators                       137522205 / 1546275507                          116 / 2383                                     
org.apache.spark.sql.catalyst.optimizer.NullPropagation                           109179965 / 1382088206                          59 / 1611                                      
org.apache.spark.sql.catalyst.optimizer.OptimizeJsonExprs                         0 / 1326862595                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.SimplifyBinaryComparison                  413782 / 1314356052                             1 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.SimplifyConditionals                      424467 / 1302138752                             1 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.CollapseProject                           220138136 / 1297546216                          220 / 1997                                     
org.apache.spark.sql.catalyst.optimizer.ConstantFolding                           204135630 / 1285454678                          197 / 1611                                     
org.apache.spark.sql.catalyst.optimizer.UnwrapCastInBinaryComparison              12264430 / 1278802169                           8 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.LikeSimplification                        677386 / 1269174316                             1 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.SimplifyCaseConversionExpressions         0 / 1261731288                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.InferFiltersFromConstraints               1187921660 / 1251528301                         278 / 386                                      
org.apache.spark.sql.catalyst.optimizer.SimplifyExtractValueOps                   0 / 1247930226                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.PushFoldableIntoBranches                  15151058 / 1247174639                           19 / 1611                                      
org.apache.spark.sql.catalyst.optimizer.SimplifyCasts                             100866650 / 1236069244                          83 / 1611                                      
org.apache.spark.sql.catalyst.optimizer.ReorderAssociativeOperator                0 / 1232823764                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.OptimizeIn                                19230987 / 1209426575                           27 / 1611                                      
org.apache.spark.sql.catalyst.optimizer.RemoveDispensableExpressions              0 / 1200696182                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.ReplaceNullWithFalseInPredicate           12294169 / 1187601942                           17 / 1611                                      
org.apache.spark.sql.catalyst.optimizer.CombineUnions                             45074167 / 1151074096                           43 / 1997                                      
org.apache.spark.sql.catalyst.optimizer.CollapseRepartition                       0 / 1111211134                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.ConstantPropagation                       0 / 1051772192                                  0 / 1611                                       
org.apache.spark.sql.catalyst.analysis.UpdateAttributeNullability                 19068252 / 1026787123                           12 / 1453                                      
org.apache.spark.sql.catalyst.optimizer.CombineFilters                            0 / 1026089915                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.CollapseWindow                            0 / 1019354027                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.EliminateSerialization                    0 / 1006513732                                  0 / 1611                                       
org.apache.spark.sql.catalyst.optimizer.SimplifyConditionalInPredicate            2930801 / 1004655515                            6 / 1611                                       

```

**Comment by wangyum** [comment]
It seems we need to add a new rule, this is because we can not add it to `ReplaceNullWithFalseInPredicate ` or `SimplifyConditionals`, example:
`select if(null, true, false)` can not rewrite to `select null and true`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.apache.spark.sql.catalyst.expressions.{And, CaseWhen, Expression, If, Literal, Not, Or}
+import org.apache.spark.sql.catalyst.expressions.Literal.{FalseLiteral, TrueLiteral}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types.BooleanType
+import org.apache.spark.util.Utils
+
+
+object SimplifyConditionalsInPredicate extends Rule[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:28]: can we add some comments to explain this rule?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    case If(cond, t, TrueLiteral) => Or(Not(cond), t)
+    case If(cond, FalseLiteral, f) => And(Not(cond), f)
+    case If(cond, TrueLiteral, f) => Or(cond, f)
+    case CaseWhen(Seq((cond, trueValue)),
+        Some(FalseLiteral) | Some(Literal(null, BooleanType)) | None) =>
+      And(cond, trueValue)
+    case CaseWhen(Seq((cond, trueValue)), Some(TrueLiteral)) =>
+      Or(Not(cond), trueValue)
+    case CaseWhen(Seq((cond, FalseLiteral)), elseValue) =>
+      And(Not(cond), elseValue.getOrElse(Literal(null, BooleanType)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:51]: `elseValue.getOrElse(FalseLiteral)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    case CaseWhen(Seq((cond, TrueLiteral)), elseValue) =>
+      Or(cond, elseValue.getOrElse(Literal(null, BooleanType)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:53]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    case e if e.dataType == BooleanType => e
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:54]: We can add an assert. The analyzer should guarantee it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    case If(cond, trueValue, TrueLiteral) => Or(Not(cond), trueValue)
+    case If(cond, FalseLiteral, falseValue) => And(Not(cond), falseValue)
+    case If(cond, TrueLiteral, falseValue) => Or(cond, falseValue)
+    case CaseWhen(Seq((cond, trueValue)),
+        Some(FalseLiteral) | Some(Literal(null, BooleanType)) | None) =>
+      And(cond, trueValue)
+    case CaseWhen(Seq((cond, trueValue)), Some(TrueLiteral)) =>
+      Or(Not(cond), trueValue)
+    case CaseWhen(Seq((cond, FalseLiteral)), elseValue) =>
+      And(Not(cond), elseValue.getOrElse(FalseLiteral))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:68]: We can further optimize it to:
```scala
    case CaseWhen(Seq((cond, FalseLiteral)), elseValue) =>
      elseValue.map(And(Not(cond), _)).getOrElse(FalseLiteral)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    case CaseWhen(Seq((cond, TrueLiteral)), elseValue) =>
+      Or(cond, elseValue.getOrElse(FalseLiteral))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:70]: We can further optimize it to:
```scala
    case CaseWhen(Seq((cond, TrueLiteral)), elseValue) =>
      elseValue.map(Or(cond, _)).getOrElse(cond)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{And, CaseWhen, Expression, If, Literal, Not, Or}
+import org.apache.spark.sql.catalyst.expressions.Literal.{FalseLiteral, TrueLiteral}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types.BooleanType
+
+/**
+ * A rule that converting conditional expressions to predicate expressions, if possible, in the
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:27]: `that converting` -> `that converts`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:75]: I mean something like
```
case e =>
  assert(e.dataType != BooleanType, ...)
  e
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum and @cloud-fan .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala (hunk 5 above)
Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:75]: It seems the added assert can never be triggered.
I think @cloud-fan meant also to remove the `case e if e.dataType == BooleanType => e`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    case If(cond, trueValue, FalseLiteral) => And(cond, trueValue)
+    case If(cond, trueValue, TrueLiteral) => Or(Not(cond), trueValue)
+    case If(cond, FalseLiteral, falseValue) => And(Not(cond), falseValue)
+    case If(cond, TrueLiteral, falseValue) => Or(cond, falseValue)
+    case CaseWhen(Seq((cond, trueValue)),
+        Some(FalseLiteral) | Some(Literal(null, BooleanType)) | None) =>
+      And(cond, trueValue)
+    case CaseWhen(Seq((cond, trueValue)), Some(TrueLiteral)) =>
+      Or(Not(cond), trueValue)
+    case CaseWhen(Seq((_, FalseLiteral)), Some(FalseLiteral) | None) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:67]: This skips evaluating the condition, and we should make sure the condition is deterministic. @wangyum can you fix it?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:67]: OK, will fix it later.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:67]: and also it matters if `cond` throws an exception. Should probably think about `NoThrow` too.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:67]: Do we need to make these condition deterministic?
https://github.com/apache/spark/blob/26d8df300a1a57e220b1a0f9814795f68101f28b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L295-L298

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:67]: https://github.com/apache/spark/pull/31067

**Comment by bart-samwel** [comment]
I'm worried about this change. In ANSI dialect, conditions can easily throw exceptions, and this change could cause _more_ conditions to be evaluated. E.g. `IF(a, b, FALSE)` turns into `a AND b`. But what if `b` can throw an error? Multiple things can happen:

If `b` can be pushed down but `a` cannot, then the shortcut evaluation behavior of AND is lost, and `b` is evaluated for values where `a` is not true. That is a semantic regression, because `IF` can be used to guard against such errors. An example case would be:

```
SELECT ...
FROM stores s JOIN store_revenues r ON s.id = r.store_id
WHERE IF(s.has_customers, r.revenue / r.num_customers > 0, FALSE)
```

In this case, the division filter can be pushed down into the scan of `store_revenues`, but it will cause division by zero.

Another case is where we are able to push down *both* `a` and `b`, and somehow the conjunctions get reordered in the process. E.g. `IF(r.num_customers > 0,  r.revenue / r.num_customers > 0, FALSE)` could turn into `r.revenue / r.num_customers > 0 AND r.num_customers > 0`, doing the division by zero first. This reordering is not possible with IF, because that expression semantically enforces the ordering.

**Comment by cloud-fan** [comment]
Yea, I think we need to take care of the "may fail" property when optimizing expressions, as well as the "deterministic" property. Many optimizer rules need to be revisited.

**Comment by shardulm94** [comment]
@cloud-fan Was there any followup work done to address the issue pointed out by @bart-samwel where some expressions may fail is evaluated without the condition being evaluated first? Seems like converting `IF` to `AND` is not safe.

**Comment by cloud-fan** [comment]
conditional expressions is the only way people can control side effects. e.g. `IF(cond, trueExpr, falseExpr)` guarantees that `trueExpr` will only be evaluated if condition is true. This optimization breaks it and the benefits do not seem significant.

@wangyum shall we revert this optimizer rule? cc @viirya @rednaxelafx @sigmod

**Comment by wangyum** [comment]
OK. revert pr: https://github.com/apache/spark/pull/37729

**Pull request closed without merging**